PEP 630: add disclaimers re. heap types & conversion, add PyType_GetModuleByDef (#2319)
python/peps: pep-0630.rst

   that shares any Python objects across interpreters implicitly depends
   on CPython's current, process-wide GIL.

Because they are immutable and process-global, static types cannot access
“their” module state.
If any method of such a type requires access to module state,
the type must be converted to a *heap-allocated type*, or *heap type*
for short. These correspond more closely to classes created by Python’s
``class`` statement.

For new modules, using heap types by default is a good rule of thumb.

> Existing classes can be converted to heap ones, but note that

Reviewer: ```suggestion
Static types can be converted to heap types, but note that
```

> the heap type API was not designed for “lossless” conversion

Reviewer: "Lossless" is a little vague IMO. Could you reword this with a clearer language?

> from static types. Unlike static types, heap type objects are mutable.

Reviewer: Heap type objects are _mutable by default_. With the current wording it sounds like they are always mutable; that is not true. I strongly suggest you reconsider your wording, in order to avoid misunderstandings. Explicit is better than implicit.

```suggestion
from static types. Unlike static types, heap type objects are mutable by default.
Heap types can be made immutable by using the flag ``Py_TPFLAGS_IMMUTABLETYPE``.
```

Author: Thanks! Somehow I forgot about that 3.10 addition of Py_TPFLAGS_IMMUTABLETYPE. 
Adding "by default" makes the PEP technically correct, though it'll be a while before common libraries can drop 3.9 and use immutable heap types.
Luckily, having types immutable is a lot more important in stdlib than elsewhere.

> the heap type API was not designed for “lossless” conversion
> from static types. Unlike static types, heap type objects are mutable.

Author: ```suggestion
the heap type API was not designed for “lossless” conversion
from static types -- that is, creating a type that works exactly like a given
static type. Unlike static types, heap type objects are mutable by default.
```

Also, when rewriting the class definition in a new API,

> you are likely to unintentionally change a few details (e.g. pickle-ability

Reviewer: I would reconsider this wording. _Take care not to_ IMO sounds nicer than _You are likely to_.

Reviewer: Also, is the pickle issue still an issue? Wasn't that fixed by Serhiy long ago? The bpo issue that mentioned this issue is closed, marked as resolved.

Author: To get back to the original comment, I think changing the behavior is a fact of life rather than something to avoid. At least right now.
(It's different in the stdlib.)

Reviewer: > OTOH, there are unexplored issues that stdlib doesn't run into but other libraries do, e.g. ones involving complex inheritance chains or metaclasses.
> 
> To keep the PEP general, I don't want to include the specific fixes here.

Sounds good to me.

For a method to get its “defining class”, it must use the

> the ``METH_METHOD | METH_FASTCALL | METH_KEYWORDS`` `calling

Reviewer: A spurious grammatical article snuck in here.

> the ``METH_METHOD | METH_FASTCALL | METH_KEYWORDS`` `calling

Author: ```
METH_METHOD | METH_FASTCALL | METH_KEYWORDS`` `calling
```

       {NULL},
   }

Module State Access from Slot Methods, Getters and Setters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note::

   This is new in Python 3.11.

>    .. After adding to limited API:

Reviewer 2: This block is commented out, looks like that's not intentional.

Author: It is. The function is not in the limited API yet. I wrote the block before the decision to not include it, and didn't want todelete it :)

      If you use the `limited API <https://docs.python.org/3/c-api/stable.html>__,
      you must update ``Py_LIMITED_API`` to ``0x030b0000``, losing ABI
      compatibility with earlier versions.

Slot methods -- the fast C equivalents for special methods, such as
`nb_add <https://docs.python.org/3/c-api/typeobj.html#c.PyNumberMethods.nb_add>`__
for ``__add__`` or `tp_new <https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_new>`__
for initialization -- have a very simple API that doesn't allow

> passsing in the defining class as in ``PyCMethod``.

Reviewer 2: ```suggestion
passing in the defining class as in ``PyCMethod``.
```

is complicated by the fact that several classes in an inheritance
hierarchy may need to reserve some state.

Lossless conversion to heap types
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The heap type API was not designed for “lossless” conversion from static types.

Reviewer: Could you reword this with a clearer language?

Author: ```suggestion
Lossless conversion to heap types
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The heap type API was not designed for “lossless” conversion from static types,
that is creating a type that works exactly like a given static type.
```
